herbert_core/utilities/geometry/get_geometry.m

```diff
@@ -0,0 +1,179 @@
+% indexes)
+[ix,iy] = ind2sub(size(ind),nodes);
+nodes = arrayfun(@(x,y)([x,y]),ix,iy,'UniformOutput',false);
+
+nb_pairs = [1,0;0,1;-1,0;0,-1];
+% 
+% Allocate memory for all nodes candidates The number of nodes is 
+% 4 = n_nodes*n_edges_from_node/2  but the memory is twice of that to
+% accound for duplicates
+edges_ind = zeros(2,8);
+ic = 1;
+for ii=1:4
+    [i0,j0] = ind2sub(size(ind),ii);
+    for jj=1:size(nb_pairs,1)
+        i1 = i0+nb_pairs(jj,1);
+        if i1<1 || i1>2; continue; end
+        j1 = j0+nb_pairs(jj,2);
+        if j1<1 || j1>2; continue; end
+        edges_ind(:,ic) = sort([ii;ind(i1,j1)]);
+        ic = ic+1;
+    end
+end
+uni_num = edges_ind(1,:)+4*edges_ind(2,:);
+[~,unii] = unique(uni_num);
+edges_ind = edges_ind(:,unii);
+
+function [nodes,edges_ind]=build3D()
+% build generic 3D geometry representation (nodes indexes and edges
+% indexes) The detailed description of the results is given in
+% get_geometry
+%
+ind = zeros(2,2,2);
+ND = numel(ind);
+%
+% number nodes linearly according to their positions
+i = 1:ND;
+ind(i) = i;
+nodes = reshape(ind,ND,1);
+% and convert node numbers into index respresentation (cellarray of node
+% indexes)
+[ix,iy,iz] = ind2sub(size(ind),nodes);
+nodes = arrayfun(@(x,y,z)([x,y,z]),ix,iy,iz,'UniformOutput',false);
+
+nb_pairs = [1,0,0; 0, 1,0; 0,0, 1;
+    -1,0,0; 0,-1,0; 0,0,-1];
+
+% Allocate memory for all nodes candidates The number of nodes is 
+% 12=n_nodes*n_edges_from_node/2  but the memory is twice of that to
+% accound for duplicates
+edges_ind = zeros(2,12*2);
+ic = 1;
+for ii=1:ND
+    [i0,j0,k0] = ind2sub(size(ind),ii);
+    for jj=1:size(nb_pairs,1)
+        i1 = i0+nb_pairs(jj,1);
+        if i1<1 || i1>2; continue; end
+        j1 = j0+nb_pairs(jj,2);
+        if j1<1 || j1>2; continue; end
+        k1 = k0+nb_pairs(jj,3);
+        if k1<1 || k1>2; continue; end
+        edges_ind(:,ic) = sort([ind(ii);ind(i1,j1,k1)]);
+        ic = ic+1;
+    end
+end
+uni_num = edges_ind(1,:)+ND*edges_ind(2,:);
+[~,unii] = unique(uni_num);
+edges_ind = edges_ind(:,unii);
+
+
+function [nodes,edges_ind]=build4D()
+% build generic 4D geometry representation (nodes indexes and edges
+% indexes) The detailed description of the results is given in
+% get_geometry.
+%
+ind = zeros(2,2,2,2);
+ND = numel(ind); % 2^4
+% number nodes linearly according to their positions
+i = 1:ND;
+ind(i) = i;
+nodes = reshape(ind,ND,1);
+%
+% and convert node numbers into index respresentation (cellarray of node
+% indexes)
+[ix,iy,iz,it] = ind2sub(size(ind),nodes);
+nodes = arrayfun(@(x,y,z,t)([x,y,z,t]),ix,iy,iz,it,'UniformOutput',false);
+
+nb_pairs = ...
+    [1,0,0,0; 0, 1,0,0;  0,0, 1,0; 0,0,0, 1;...
+    -1,0,0,0; 0,-1,0,0;  0,0,-1,0; 0,0,0,-1];
+
+% Allocate memory for all nodes candidates The number of nodes is 
+% 32 = n_nodes*n_edges_from_node/2  but the memory is twice of that to
+% accound for duplicates
+edges_ind = zeros(2,32*2);
+ic = 1;
```

Comment:
@abuts - apologies, I read 32**2 instead of 32*2, so it's not 1024, it's 64. Still wondering why you made the second argument an expression rather than a simple constant. Thanks

Reply:
its to say in expressions the same as said in sentence above. its twice of the requested number. As it causes so much confusion, I have changed it to 64